server/docs/design/block-verification.md

```diff
@@ -0,0 +1,198 @@
+  BlockHashingSession createSession(List<BlockItem> initialBlockItems, ExecutorService executorService, SignatureVerifier signatureVerifier);
+}
+```
+### BlockHashingSession
+```java
+/* Once hashing is completed internally, calls SignatureVerifier to continue with verification process */
+public interface BlockHashingSession {
+  void addBlockItem(BlockItem item);
+  CompletableFuture<Void> completeHashing(); // triggers final hash computation asynchronously  
+}
+```
+### SignatureVerifier
+```java
+public interface SignatureVerifier {
+  void verifySignature(byte[] signature, byte[] computedHash, long blockNumber);  
+}
+```
+### BlockStatusManager
+```java
+
+public interface BlockStatusManager {
+    void updateBlockStatus(long blockNumber, BlockVerificationStatus status, VerificationError error);
+}
+```
+
+## Enums
+```java
+public enum BlockVerificationStatus {
+  VERIFIED,
+  ERROR
+}
+
+public enum VerificationError {
+  NONE,
+  SIGNATURE_INVALID,
+  SYSTEM_ERROR
+}
+```
+
+## Metrics
+
+- **blocks_received**: Counter of the number of blocks received for verification
+- **blocks_verified**: Counter of the number of blocks verified
+- **blocks_unverified**: Counter of the number of blocks that have not been verified, is the difference between blocks_received and blocks_verified.
+- **blocks_signature_invalid**: Counter of the number of blocks with invalid signatures
+- **blocks_system_error**: Counter of the number of blocks with system errors
+- **block_verification_time**: Histogram of the time taken to verify a block, gives the node operator an idea of the time taken to verify a block.
+- 
+## Exceptions
+
+- **SYSTEM_ERROR:** Issues with node configuration or bugs. The node logs details, updates metrics, and might attempt recovery or halt.
+
+- **SIGNATURE_INVALID:** If verification fails, SIGNATURE_INVALID is used. The block is marked invalid, and the BlockStatusManager triggers error-handling routines (requesting re-sends, removing corrupted data, notifying subscribers, etc.).
+
+### Signature invalid
+If the computed hash does not match the hash implied by the signature, the block is considered tampered. It is marked invalid and appropriate recovery steps are taken.
+
+## Notes:
+In order to keep it simple the first implementation will not start work async as items come, but will wait for the block_proof to arrive in order to start work async, that way we avoid a thread waiting for items and idle, and avoid the complexity of activating the async work as items come. The tradeoff of this simplification is that we might experience a higher latency for the stream of verified blocks, but we can always improve this in the future if is needed since the design already contemplates the async work of block items as they are received.
+By using metrics we will be able to measure the latency of block verification and decide if we need to improve it by activating the async work of block items as they come.
```

Comment:
I would encourage you to start the async work up front for the following reasons:
1. We don't want to retain the whole block in memory, the hash algorithm laid out by Derek only keeps (for each of the two trees) the most recent item and up to 3 hash values, which is _much_ smaller (3-4 orders of magnitude).
1. If you use `CompletableFuture`, an input queue, a non-default fork-join work pool, and enable virtual threads it is quite easy to add asynchronous computation and the JVM will work-steal, which avoids any idle thread issues.
1. Any latency in verification is magnified and also creates risk because we cannot store a block until it's verified, and the publishers are waiting for acknowledgement that the block is verified before sending the next one.

Reply:
I actually agree with you on this one, but did the change at the request of @mattp-swirldslabs once we synced on the design after I share it on slack. but instead of changing the design, added notes to leave it as optional and the intention to optimize it at the future should the need arise. I think we should remove any latency where we can, but understand Matts view to want to keep things simple and safe when it comes to threads.

Reply:
I have modified the design doc so we keep async processing of hashes. That said, I will move this comments to a separate [discussion](https://github.com/hashgraph/hedera-block-node/discussions/400) on Module/Services System Loader and communication interface.